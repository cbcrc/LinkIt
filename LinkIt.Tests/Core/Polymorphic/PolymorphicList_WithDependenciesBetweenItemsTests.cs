using System.Collections.Generic;
using ApprovalTests.Reporters;
using LinkIt.ConfigBuilders;
using LinkIt.PublicApi;
using LinkIt.Tests.Shared;
using NUnit.Framework;
using RC.Testing;

namespace LinkIt.Tests.Polymorphic {
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class PolymorphicList_WithDependenciesBetweenItemsTests {
        private ILoadLinkProtocol _sut;

        [SetUp]
        public void SetUp() {
            var loadLinkProtocolBuilder = new LoadLinkProtocolBuilder();

            loadLinkProtocolBuilder.For<LinkedSource>()
                .PolymorphicLoadLinkForList(
                    linkedSource => linkedSource.Model.PolyLinks,
                    linkedSource => linkedSource.PolyLinks,
                    link => link.Type,
                    includes => includes
                        .Include<Image>().AsReferenceById(
                            "image",
                            link=>link.Id
                        )
                        .Include<PersonLinkedSource>().AsNestedLinkedSourceById(
                            "person",
                            link => link.Id
                        )
                );
            loadLinkProtocolBuilder.For<PersonLinkedSource>()
                .LoadLinkReferenceById(
                    linkedSource => linkedSource.Model.SummaryImageId,
                    linkedSource => linkedSource.SummaryImage
                );
            
            _sut = loadLinkProtocolBuilder.Build(() => new ReferenceLoaderStub());
        }

        [Test]
        public void LoadLink_WithDependenciesBetweenItems_ShouldLink3Items() {
            var actual = _sut.LoadLink<LinkedSource>().FromModel(
                new Model {
                    Id = "1",
                    PolyLinks = new List<Link>{
                        new Link{ Type = "image", Id = "before"},
                        new Link{ Type = "person", Id = "the-person"},
                        new Link{ Type = "image", Id = "after"}
                    }
                }
            );

            ApprovalsExt.VerifyPublicProperties(actual);
        }

        public class LinkedSource : ILinkedSource<Model> {
            public Model Model { get; set; }
            public List<object> PolyLinks { get; set; }
        }

        public class Model{
            public string Id { get; set; }
            public List<Link> PolyLinks { get; set; }
        }

        public class Link{
            public string Type { get; set; }
            public string Id { get; set; }
        }
    }
}
