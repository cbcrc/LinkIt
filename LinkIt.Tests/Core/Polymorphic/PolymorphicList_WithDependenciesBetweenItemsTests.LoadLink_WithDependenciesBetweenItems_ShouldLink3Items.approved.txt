{
  "model": {
    "id": "1",
    "polyLinks": [
      {
        "type": "image",
        "id": "before"
      },
      {
        "type": "person",
        "id": "the-person"
      },
      {
        "type": "image",
        "id": "after"
      }
    ]
  },
  "polyLinks": [
    {
      "$type": "LinkIt.Tests.Shared.Image, LinkIt.Tests",
      "id": "before",
      "alt": "alt-before"
    },
    {
      "$type": "LinkIt.Tests.Core.PersonLinkedSource, LinkIt.Tests",
      "model": {
        "id": "the-person",
        "name": "name-the-person",
        "summaryImageId": "person-img-the-person"
      },
      "summaryImage": {
        "id": "person-img-the-person",
        "alt": "alt-person-img-the-person"
      }
    },
    {
      "$type": "LinkIt.Tests.Shared.Image, LinkIt.Tests",
      "id": "after",
      "alt": "alt-after"
    }
  ]
}